swim.go

```diff
@@ -279,33 +308,201 @@ func (s *SWIM) startFailureDetector() {
 // 3. At the end of T, SWIM checks to see if ack was received from k members, and if there is no message,
 //    The member(j) is judged to be failed, so check the member(j) as suspected or delete the member(j) from memberMap.
 //
-
-func (s *SWIM) probe(member Member) {
+func (s *SWIM) probe(member Member, timer *time.Timer) {
 
 	if member.Status == Dead {
 		return
 	}
 
-	end := make(chan struct{}, 1)
-	defer close(end)
+	end := make(chan ProbeResponse, 1)
+	defer func() {
+		close(end)
+	}()
 
 	go func() {
+		err := s.ping(&member)
+		if err == nil {
+			end <- ProbeResponse{}
+			return
+		}
+		if err != ErrSendTimeout {
+			iLogger.Errorf(nil, "Error occurred when ping failed [%s]", err.Error())
+			return
+		}
 
-		// Ping to member
-		time.Sleep(1 * time.Second)
-		end <- struct{}{}
-	}()
+		err = s.indirectProbe(&member)
+		if err != nil {
+			end <- ProbeResponse{err: err}
+			return
+		}
 
-	T := time.NewTimer(time.Millisecond * time.Duration(s.config.T))
+		end <- ProbeResponse{}
+	}()
 
 	select {
-	case <-end:
-		// Ended
-		return
-	case <-T.C:
-		// Suspect the member.
-		return
+	// if timed-out, then suspect member
+	case <-timer.C:
+		iLogger.Infof(nil, "[SWIM] probe member [%s] timed out, start suspect", member.ID)
+		s.awareness.ApplyDelta(1)
+		s.suspect(&member)
+
+	// if probe ended with error then suspect member and increase Awareness
+	// otherwise just decrease Awareness score
+	case resp := <-end:
+		if resp.err != nil {
+			s.awareness.ApplyDelta(1)
+			s.suspect(&member)
+			return
+		}
+
+		s.awareness.ApplyDelta(-1)
+	}
+}
+
+// indirectProbe select k-random member from MemberMap, sends
+// indirect-ping to them. if one of them sends back Ack message
+// then indirectProbe success, otherwise failed.
+// if one of k-member successfully received ACK message, then cancel
+// k-1 member's probe
+func (s *SWIM) indirectProbe(target *Member) error {
+	wg := &sync.WaitGroup{}
+	wg.Add(s.config.K)
+
+	// with cancel we can send the signal to goroutines which share
+	// this @ctx context
+	ctx, cancel := context.WithCancel(context.Background())
+
+	done := make(chan pb.Message, s.config.K)
+
+	defer func() {
+		cancel()
+		close(done)
+	}()
+
+	kMembers := s.memberMap.SelectKRandomMemberID(s.config.K)
+	for _, m := range kMembers {
+		go func(mediator Member) {
+			defer wg.Done()
+
+			task := func() (interface{}, error) {
+				return s.indirectPing(mediator, *target)
+			}
+
+			resp := NewTaskRunner(task, ctx).Start()
+			if resp.payload == nil {
```

Comment:
Suppose there is a problem with some of the members of the indirect search k. Do not send a message to the done channel when there is a problem with payload or ok. But in the done channel, is there any problem comparing the message counter and config.K?